File: tests/python_prototype_tests/parser_tests.py
import unittest
from src.python_prototype.parser import parse
from src.python_prototype.lexer import tokenize

class TestParser(unittest.TestCase):
    def test_parse_function_definition(self):
        code = """
        def test(x, y):
            return x + y
        """
        tokens = tokenize(code)
        ast = parse(tokens)
        self.assertEqual(ast.type, "Program")
        self.assertEqual(len(ast.children), 1)
        func_def = ast.children[0]
        self.assertEqual(func_def.type, "FunctionDef")
        self.assertEqual(func_def.name, "test")
        self.assertEqual(len(func_def.params), 2)

    def test_parse_islamic_entry_point(self):
        code = """
        ﷽:
            print("Hello")
        """
        tokens = tokenize(code)
        ast = parse(tokens)
        self.assertEqual(ast.type, "Program")
        self.assertEqual(ast.children[0].type, "EntryPoint")

if __name__ == '__main__':
    unittest.main()

File: tests/rust_implementation_tests/parser_tests.rs
#[cfg(test)]
mod tests {
    use super::*;
    use crate::parser::Parser;
    use crate::lexer::{Lexer, TokenType};

    #[test]
    fn test_parse_function_definition() {
        let code = "def test(x, y): return x + y";
        let mut lexer = Lexer::new(code);
        lexer.tokenize().unwrap();
        let mut parser = Parser::new(lexer.tokens);
        let ast = parser.parse().unwrap();
        
        match ast.node_type {
            AstNodeType::Program => {
                let func_def = &ast.children[0];
                match &func_def.node_type {
                    AstNodeType::FunctionDef(name, params) => {
                        assert_eq!(name, "test");
                        assert_eq!(params, &vec!["x".to_string(), "y".to_string()]);
                    },
                    _ => panic!("Expected function definition")
                }
            },
            _ => panic!("Expected program node")
        }
    }
}

